audio/mods/mod_xm_s3m.cpp

```diff
@@ -0,0 +1,1352 @@
+			}
+		}
+	}
+}
+
+void ModXmS3mStream::updateSampleIdx(Channel &channel, int count, int sampleRate) {
+	Sample *sample = channel.sample;
+	int step = (channel.freq << (FP_SHIFT - 3)) / (sampleRate >> 3);
+	channel.sampleFra += step * count;
+	channel.sampleIdx += channel.sampleFra >> FP_SHIFT;
+	if (channel.sampleIdx > (int)sample->loopStart) {
+		if (sample->loopLength > 1) {
+			channel.sampleIdx = sample->loopStart + (channel.sampleIdx - sample->loopStart) % sample->loopLength;
+		} else {
+			channel.sampleIdx = sample->loopStart;
+		}
+	}
+	channel.sampleFra &= FP_MASK;
+}
+
+void ModXmS3mStream::volumeRamp(int *mixBuf, int tickLen) {
+	int rampRate = 256 * 2048 / _sampleRate;
+	for (int idx = 0, a1 = 0; a1 < 256; idx += 2, a1 += rampRate) {
+		int a2 = 256 - a1;
+		mixBuf[idx] = (mixBuf[idx] * a1 + _rampBuf[idx] * a2) >> 8;
+		mixBuf[idx + 1] = (mixBuf[idx + 1] * a1 + _rampBuf[idx + 1] * a2) >> 8;
+	}
+	memcpy(_rampBuf, &mixBuf[tickLen * 2], 128 * sizeof(int));
+}
+
+/* 2:1 downsampling with simple but effective anti-aliasing. Buf must contain count * 2 + 1 stereo samples. */
+void ModXmS3mStream::downsample(int *buf, int count) {
+	int outLen = count * 2;
+	for (int idx = 0, outIdx = 0; outIdx < outLen; idx += 4, outIdx += 2) {
+		buf[outIdx] = (buf[idx] >> 2) + (buf[idx + 2] >> 1) + (buf[idx + 4] >> 2);
+		buf[outIdx + 1] = (buf[idx + 1] >> 2) + (buf[idx + 3] >> 1) + (buf[idx + 5] >> 2);
+	}
+}
+
+/* Generates audio and returns the number of stereo samples written into mixBuf. */
+int ModXmS3mStream::getAudio(int *mixBuf) {
+	if (_mixBuffer) {
+		memcpy(mixBuf, _mixBuffer, _mixBufferSamples * sizeof(int));
+		delete []_mixBuffer;
+		_mixBuffer = nullptr;
+		return _mixBufferSamples;
+	}
+
+	int tickLen = calculateTickLength();
+	/* Clear output buffer. */
+	memset(mixBuf, 0, (tickLen + 65) * 4 * sizeof(int));
+	/* Resample. */
+	for (int idx = 0; idx < _module.numChannels; idx++) {
+		Channel &channel = _channels[idx];
+		resample(channel, mixBuf, 0, (tickLen + 65) * 2, _sampleRate * 2);
+		updateSampleIdx(channel, tickLen * 2, _sampleRate * 2);
+	}
+	downsample(mixBuf, tickLen + 64);
+	volumeRamp(mixBuf, tickLen);
+	tick();
+	return tickLen * 2;  // stereo samples
+}
+
+int ModXmS3mStream::readBuffer(int16 *buffer, const int numSamples) {
+	int samplesRead = 0;
+	int8 *ptr = (int8 *)buffer;
+	while (samplesRead < numSamples && _dataLeft >= 0) {
+		int *mixBuf = new int[calculateMixBufLength()];
+		int samples = getAudio(mixBuf);
+		if (samplesRead + samples > numSamples) {
+			_mixBufferSamples = samplesRead + samples - numSamples;
+			samples -= _mixBufferSamples;
+			_mixBuffer = new int[_mixBufferSamples];
+			memcpy(_mixBuffer, mixBuf + samples, _mixBufferSamples * sizeof(int));
+		}
+		for (int idx = 0; idx < samples; ++idx) {
+			int ampl = mixBuf[idx];
+			if (ampl > 32767) {
+				ampl = 32767;
+			}
+			if (ampl < -32768) {
+				ampl = -32768;
+			}
+			*ptr++ = ampl & 0xFF;
+			*ptr++ = (ampl >> 8) & 0xFF;
```

Comment:
The AudioStream documentation says the samples must be in native endianness. So you should just copy the 16 bits sample to the output buffer without explicitly storing the bytes in little endian order.

Reply:
Ok, I see